// relude/Relude_Debounce.js
// Generated by Melange
'use strict';

var Curry = require("melange.runtime/curry.js");
var Caml_option = require("melange.runtime/caml_option.js");
var Relude__Relude_Option = require("./Relude_Option.js");

function debounce(delayMS, leadingOpt, f) {
  var leading = leadingOpt !== undefined ? leadingOpt : false;
  var timerId = {
    contents: undefined
  };
  var cancel = function (param) {
    Curry._2(Relude__Relude_Option.forEach, (function (timerId) {
            clearTimeout(timerId);
          }), timerId.contents);
    timerId.contents = undefined;
  };
  var isScheduled = function (param) {
    return Relude__Relude_Option.isSome(timerId.contents);
  };
  var flush = function (param) {
    cancel(undefined);
    Curry._1(f, undefined);
  };
  var doLeading = {
    contents: leading
  };
  var debounced = function (param) {
    if (doLeading.contents) {
      doLeading.contents = false;
      return Curry._1(f, undefined);
    } else {
      cancel(undefined);
      timerId.contents = Caml_option.some(setTimeout((function (param) {
                  Curry._1(f, undefined);
                  timerId.contents = undefined;
                }), delayMS));
      return ;
    }
  };
  return {
          cancel: cancel,
          flush: flush,
          isScheduled: isScheduled,
          f: debounced
        };
}

exports.debounce = debounce;
/* Relude__Relude_Option Not a pure module */
