// relude/Relude_Tuple.js
// Generated by Melange
'use strict';

var Relude__Relude_Tuple2 = require("./Relude_Tuple2.js");
var Relude__Relude_Tuple3 = require("./Relude_Tuple3.js");
var Relude__Relude_Tuple4 = require("./Relude_Tuple4.js");
var Relude__Relude_Tuple5 = require("./Relude_Tuple5.js");
var Relude__Relude_Function = require("./Relude_Function.js");

var make = Relude__Relude_Tuple2.make;

var make2 = Relude__Relude_Tuple2.make;

var make3 = Relude__Relude_Tuple3.make;

var make4 = Relude__Relude_Tuple4.make;

var make5 = Relude__Relude_Tuple5.make;

var fromArray = Relude__Relude_Tuple2.fromArray;

var fromArray2 = Relude__Relude_Tuple2.fromArray;

var fromArray3 = Relude__Relude_Tuple3.fromArray;

var fromArray4 = Relude__Relude_Tuple4.fromArray;

var fromArray5 = Relude__Relude_Tuple5.fromArray;

var fromArrayAtLeast = Relude__Relude_Tuple2.fromArrayAtLeast;

var fromArrayAtLeast2 = Relude__Relude_Tuple2.fromArrayAtLeast;

var fromArrayAtLeast3 = Relude__Relude_Tuple3.fromArrayAtLeast;

var fromArrayAtLeast4 = Relude__Relude_Tuple4.fromArrayAtLeast;

var fromArrayAtLeast5 = Relude__Relude_Tuple5.fromArrayAtLeast;

var fromList = Relude__Relude_Tuple2.fromList;

var fromList2 = Relude__Relude_Tuple2.fromList;

var fromList3 = Relude__Relude_Tuple3.fromList;

var fromList4 = Relude__Relude_Tuple4.fromList;

var fromList5 = Relude__Relude_Tuple5.fromList;

var fromListAtLeast = Relude__Relude_Tuple2.fromListAtLeast;

var fromListAtLeast2 = Relude__Relude_Tuple2.fromListAtLeast;

var fromListAtLeast3 = Relude__Relude_Tuple3.fromListAtLeast;

var fromListAtLeast4 = Relude__Relude_Tuple4.fromListAtLeast;

var fromListAtLeast5 = Relude__Relude_Tuple5.fromListAtLeast;

var apply = Relude__Relude_Function.uncurry2;

var apply2 = Relude__Relude_Function.uncurry2;

var apply3 = Relude__Relude_Function.uncurry3;

var apply4 = Relude__Relude_Function.uncurry4;

var apply5 = Relude__Relude_Function.uncurry5;

var first = Relude__Relude_Tuple2.first;

var first2 = Relude__Relude_Tuple2.first;

var first3 = Relude__Relude_Tuple3.first;

var first4 = Relude__Relude_Tuple4.first;

var first5 = Relude__Relude_Tuple5.first;

var second = Relude__Relude_Tuple2.second;

var second2 = Relude__Relude_Tuple2.second;

var second3 = Relude__Relude_Tuple3.second;

var second4 = Relude__Relude_Tuple4.second;

var second5 = Relude__Relude_Tuple5.second;

var third = Relude__Relude_Tuple3.third;

var third3 = Relude__Relude_Tuple3.third;

var third4 = Relude__Relude_Tuple4.third;

var third5 = Relude__Relude_Tuple5.third;

var fourth = Relude__Relude_Tuple4.fourth;

var fourth4 = Relude__Relude_Tuple4.fourth;

var fourth5 = Relude__Relude_Tuple5.fourth;

var fifth = Relude__Relude_Tuple5.fifth;

var fifth5 = Relude__Relude_Tuple5.fifth;

var showBy = Relude__Relude_Tuple2.showBy;

var showBy2 = Relude__Relude_Tuple2.showBy;

var showBy3 = Relude__Relude_Tuple3.showBy;

var showBy4 = Relude__Relude_Tuple4.showBy;

var showBy5 = Relude__Relude_Tuple5.showBy;

var eqBy = Relude__Relude_Tuple2.eqBy;

var eqBy2 = Relude__Relude_Tuple2.eqBy;

var eqBy3 = Relude__Relude_Tuple3.eqBy;

var eqBy4 = Relude__Relude_Tuple4.eqBy;

var eqBy5 = Relude__Relude_Tuple5.eqBy;

var compareBy = Relude__Relude_Tuple2.compareBy;

var compareBy2 = Relude__Relude_Tuple2.compareBy;

var compareBy3 = Relude__Relude_Tuple3.compareBy;

var compareBy4 = Relude__Relude_Tuple4.compareBy;

var compareBy5 = Relude__Relude_Tuple5.compareBy;

exports.make = make;
exports.make2 = make2;
exports.make3 = make3;
exports.make4 = make4;
exports.make5 = make5;
exports.fromArray = fromArray;
exports.fromArray2 = fromArray2;
exports.fromArray3 = fromArray3;
exports.fromArray4 = fromArray4;
exports.fromArray5 = fromArray5;
exports.fromArrayAtLeast = fromArrayAtLeast;
exports.fromArrayAtLeast2 = fromArrayAtLeast2;
exports.fromArrayAtLeast3 = fromArrayAtLeast3;
exports.fromArrayAtLeast4 = fromArrayAtLeast4;
exports.fromArrayAtLeast5 = fromArrayAtLeast5;
exports.fromList = fromList;
exports.fromList2 = fromList2;
exports.fromList3 = fromList3;
exports.fromList4 = fromList4;
exports.fromList5 = fromList5;
exports.fromListAtLeast = fromListAtLeast;
exports.fromListAtLeast2 = fromListAtLeast2;
exports.fromListAtLeast3 = fromListAtLeast3;
exports.fromListAtLeast4 = fromListAtLeast4;
exports.fromListAtLeast5 = fromListAtLeast5;
exports.apply = apply;
exports.apply2 = apply2;
exports.apply3 = apply3;
exports.apply4 = apply4;
exports.apply5 = apply5;
exports.first = first;
exports.first2 = first2;
exports.first3 = first3;
exports.first4 = first4;
exports.first5 = first5;
exports.second = second;
exports.second2 = second2;
exports.second3 = second3;
exports.second4 = second4;
exports.second5 = second5;
exports.third = third;
exports.third3 = third3;
exports.third4 = third4;
exports.third5 = third5;
exports.fourth = fourth;
exports.fourth4 = fourth4;
exports.fourth5 = fourth5;
exports.fifth = fifth;
exports.fifth5 = fifth5;
exports.showBy = showBy;
exports.showBy2 = showBy2;
exports.showBy3 = showBy3;
exports.showBy4 = showBy4;
exports.showBy5 = showBy5;
exports.eqBy = eqBy;
exports.eqBy2 = eqBy2;
exports.eqBy3 = eqBy3;
exports.eqBy4 = eqBy4;
exports.eqBy5 = eqBy5;
exports.compareBy = compareBy;
exports.compareBy2 = compareBy2;
exports.compareBy3 = compareBy3;
exports.compareBy4 = compareBy4;
exports.compareBy5 = compareBy5;
/* Relude__Relude_Tuple2 Not a pure module */
