// relude/Relude_StringMap.js
// Generated by Melange
'use strict';

var Relude__Relude_Map = require("./Relude_Map.js");
var Relude__Relude_String = require("./Relude_String.js");

var include = Relude__Relude_Map.WithOrd(Relude__Relude_String.Ord);

var Comparable = include.Comparable;

var make = include.make;

var set = include.set;

var singleton = include.singleton;

var isEmpty = include.isEmpty;

var contains = include.contains;

var compareInt = include.compareInt;

var compareBy = include.compareBy;

var eqBy = include.eqBy;

var find = include.find;

var forEach = include.forEach;

var foldLeft = include.foldLeft;

var all = include.all;

var any = include.any;

var length = include.length;

var toArray = include.toArray;

var fromArray = include.fromArray;

var fromValueArray = include.fromValueArray;

var toList = include.toList;

var fromList = include.fromList;

var fromValueList = include.fromValueList;

var keys = include.keys;

var keyArray = include.keyArray;

var values = include.values;

var valueArray = include.valueArray;

var minKey = include.minKey;

var maxKey = include.maxKey;

var min = include.min;

var max = include.max;

var get = include.get;

var getOrElse = include.getOrElse;

var remove = include.remove;

var removeMany = include.removeMany;

var update = include.update;

var merge = include.merge;

var mergeMany = include.mergeMany;

var filter = include.filter;

var keep = include.keep;

var filterNot = include.filterNot;

var reject = include.reject;

var partition = include.partition;

var map = include.map;

var mapWithKey = include.mapWithKey;

var groupListBy = include.groupListBy;

var groupArrayBy = include.groupArrayBy;

exports.Comparable = Comparable;
exports.make = make;
exports.set = set;
exports.singleton = singleton;
exports.isEmpty = isEmpty;
exports.contains = contains;
exports.compareInt = compareInt;
exports.compareBy = compareBy;
exports.eqBy = eqBy;
exports.find = find;
exports.forEach = forEach;
exports.foldLeft = foldLeft;
exports.all = all;
exports.any = any;
exports.length = length;
exports.toArray = toArray;
exports.fromArray = fromArray;
exports.fromValueArray = fromValueArray;
exports.toList = toList;
exports.fromList = fromList;
exports.fromValueList = fromValueList;
exports.keys = keys;
exports.keyArray = keyArray;
exports.values = values;
exports.valueArray = valueArray;
exports.minKey = minKey;
exports.maxKey = maxKey;
exports.min = min;
exports.max = max;
exports.get = get;
exports.getOrElse = getOrElse;
exports.remove = remove;
exports.removeMany = removeMany;
exports.update = update;
exports.merge = merge;
exports.mergeMany = mergeMany;
exports.filter = filter;
exports.keep = keep;
exports.filterNot = filterNot;
exports.reject = reject;
exports.partition = partition;
exports.map = map;
exports.mapWithKey = mapWithKey;
exports.groupListBy = groupListBy;
exports.groupArrayBy = groupArrayBy;
/* include Not a pure module */
