// relude/Relude_ReaderT.js
// Generated by Melange
'use strict';

var Curry = require("melange.runtime/curry.js");
var Relude__Relude_Function = require("./Relude_Function.js");
var Relude__Relude_Identity = require("./Relude_Identity.js");
var Relude__Relude_Extensions_Apply = require("./extensions/Relude_Extensions_Apply.js");
var Relude__Relude_Extensions_Monad = require("./extensions/Relude_Extensions_Monad.js");
var Relude__Relude_Extensions_Functor = require("./extensions/Relude_Extensions_Functor.js");
var Relude__Relude_Extensions_Applicative = require("./extensions/Relude_Extensions_Applicative.js");

function WithMonad(M) {
  var make = function (rToMA) {
    return /* ReaderT */{
            _0: rToMA
          };
  };
  var runReaderT = function (r, rToMA) {
    return Curry._1(rToMA._0, r);
  };
  var mapReaderT = function (maToMB, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (param) {
                return Relude__Relude_Function.Infix.$less$less(maToMB, rToMA$1, param);
              })
          };
  };
  var withReaderT = function (r2ToR1, r1ToMA) {
    var r1ToMA$1 = r1ToMA._0;
    return /* ReaderT */{
            _0: (function (param) {
                return Relude__Relude_Function.Infix.$less$less(r1ToMA$1, r2ToR1, param);
              })
          };
  };
  var ask = /* ReaderT */{
    _0: (function (r) {
        return Curry._1(M.pure, r);
      })
  };
  var asks = function (rToA) {
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._1(M.pure, Curry._1(rToA, r));
              })
          };
  };
  var local = function (rToR, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._1(rToMA$1, Curry._1(rToR, r));
              })
          };
  };
  var map = function (aToB, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.map, aToB, Curry._1(rToMA$1, r));
              })
          };
  };
  var apply = function (rToMAToB, rToMA) {
    var rToMA$1 = rToMA._0;
    var rToMAToB$1 = rToMAToB._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.apply, Curry._1(rToMAToB$1, r), Curry._1(rToMA$1, r));
              })
          };
  };
  var pure = function (a) {
    return /* ReaderT */{
            _0: (function (param) {
                return Curry._1(M.pure, a);
              })
          };
  };
  var bind = function (rToMA, aToReaderB) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.flat_map, Curry._1(rToMA$1, r), (function (a) {
                              var rToMB = Curry._1(aToReaderB, a);
                              return Curry._1(rToMB._0, r);
                            }));
              })
          };
  };
  var semiflatMap = function (aToMA, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.flat_map, Curry._1(rToMA$1, r), Curry.__1(aToMA));
              })
          };
  };
  var WithEnv = function (R) {
    var Functor = {
      map: map
    };
    var include = Relude__Relude_Extensions_Functor.FunctorExtensions(Functor);
    var Apply = {
      map: map,
      apply: apply
    };
    var include$1 = Relude__Relude_Extensions_Apply.ApplyExtensions(Apply);
    var Applicative = {
      map: map,
      apply: apply,
      pure: pure
    };
    var include$2 = Relude__Relude_Extensions_Applicative.ApplicativeExtensions(Applicative);
    var Monad = {
      map: map,
      apply: apply,
      pure: pure,
      flat_map: bind
    };
    var include$3 = Relude__Relude_Extensions_Monad.MonadExtensions(Monad);
    var include$4 = Relude__Relude_Extensions_Functor.FunctorInfix(Functor);
    var include$5 = Relude__Relude_Extensions_Apply.ApplyInfix(Apply);
    var include$6 = Relude__Relude_Extensions_Monad.MonadInfix(Monad);
    var Infix_FunctorExtensions = include$4.FunctorExtensions;
    var Infix_$less$$great = include$4.$less$$great;
    var Infix_$less$$$great = include$4.$less$$$great;
    var Infix_$less$ = include$4.$less$;
    var Infix_$$great = include$4.$$great;
    var Infix_$less$at$great = include$4.$less$at$great;
    var Infix_ApplyExtensions = include$5.ApplyExtensions;
    var Infix_$less$star$great = include$5.$less$star$great;
    var Infix_$less$star = include$5.$less$star;
    var Infix_$star$great = include$5.$star$great;
    var Infix_MonadExtensions = include$6.MonadExtensions;
    var Infix_$great$great$eq = include$6.$great$great$eq;
    var Infix_$eq$less$less = include$6.$eq$less$less;
    var Infix_$great$eq$great = include$6.$great$eq$great;
    var Infix_$less$eq$less = include$6.$less$eq$less;
    var Infix = {
      FunctorExtensions: Infix_FunctorExtensions,
      $less$$great: Infix_$less$$great,
      $less$$$great: Infix_$less$$$great,
      $less$: Infix_$less$,
      $$great: Infix_$$great,
      $less$at$great: Infix_$less$at$great,
      ApplyExtensions: Infix_ApplyExtensions,
      $less$star$great: Infix_$less$star$great,
      $less$star: Infix_$less$star,
      $star$great: Infix_$star$great,
      MonadExtensions: Infix_MonadExtensions,
      $great$great$eq: Infix_$great$great$eq,
      $eq$less$less: Infix_$eq$less$less,
      $great$eq$great: Infix_$great$eq$great,
      $less$eq$less: Infix_$less$eq$less
    };
    return {
            make: make,
            runReaderT: runReaderT,
            mapReaderT: mapReaderT,
            withReaderT: withReaderT,
            ask: ask,
            asks: asks,
            local: local,
            semiflatMap: semiflatMap,
            Functor: Functor,
            map: map,
            BsFunctorExtensions: include.BsFunctorExtensions,
            flipMap: include.flipMap,
            $$void: include.$$void,
            voidRight: include.voidRight,
            voidLeft: include.voidLeft,
            flap: include.flap,
            Apply: Apply,
            apply: apply,
            BsApplyExtensions: include$1.BsApplyExtensions,
            applyFirst: include$1.applyFirst,
            applySecond: include$1.applySecond,
            map2: include$1.map2,
            map3: include$1.map3,
            map4: include$1.map4,
            map5: include$1.map5,
            tuple2: include$1.tuple2,
            tuple3: include$1.tuple3,
            tuple4: include$1.tuple4,
            tuple5: include$1.tuple5,
            mapTuple2: include$1.mapTuple2,
            mapTuple3: include$1.mapTuple3,
            mapTuple4: include$1.mapTuple4,
            mapTuple5: include$1.mapTuple5,
            Applicative: Applicative,
            pure: pure,
            BsApplicativeExtensions: include$2.BsApplicativeExtensions,
            liftA1: include$2.liftA1,
            all: include$2.all,
            Monad: Monad,
            bind: bind,
            BsMonadExtensions: include$3.BsMonadExtensions,
            flatMap: include$3.flatMap,
            flatten: include$3.flatten,
            composeKleisli: include$3.composeKleisli,
            flipComposeKleisli: include$3.flipComposeKleisli,
            liftM1: include$3.liftM1,
            when_: include$3.when_,
            unless: include$3.unless,
            Infix: Infix
          };
  };
  return {
          make: make,
          runReaderT: runReaderT,
          mapReaderT: mapReaderT,
          withReaderT: withReaderT,
          ask: ask,
          asks: asks,
          local: local,
          map: map,
          apply: apply,
          pure: pure,
          bind: bind,
          semiflatMap: semiflatMap,
          WithEnv: WithEnv
        };
}

function WithMonadAndEnv(M, E) {
  var make = function (rToMA) {
    return /* ReaderT */{
            _0: rToMA
          };
  };
  var runReaderT = function (r, rToMA) {
    return Curry._1(rToMA._0, r);
  };
  var mapReaderT = function (maToMB, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (param) {
                return Relude__Relude_Function.Infix.$less$less(maToMB, rToMA$1, param);
              })
          };
  };
  var withReaderT = function (r2ToR1, r1ToMA) {
    var r1ToMA$1 = r1ToMA._0;
    return /* ReaderT */{
            _0: (function (param) {
                return Relude__Relude_Function.Infix.$less$less(r1ToMA$1, r2ToR1, param);
              })
          };
  };
  var ask = /* ReaderT */{
    _0: (function (r) {
        return Curry._1(M.pure, r);
      })
  };
  var asks = function (rToA) {
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._1(M.pure, Curry._1(rToA, r));
              })
          };
  };
  var local = function (rToR, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._1(rToMA$1, Curry._1(rToR, r));
              })
          };
  };
  var map = function (aToB, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.map, aToB, Curry._1(rToMA$1, r));
              })
          };
  };
  var apply = function (rToMAToB, rToMA) {
    var rToMA$1 = rToMA._0;
    var rToMAToB$1 = rToMAToB._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.apply, Curry._1(rToMAToB$1, r), Curry._1(rToMA$1, r));
              })
          };
  };
  var pure = function (a) {
    return /* ReaderT */{
            _0: (function (param) {
                return Curry._1(M.pure, a);
              })
          };
  };
  var bind = function (rToMA, aToReaderB) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.flat_map, Curry._1(rToMA$1, r), (function (a) {
                              var rToMB = Curry._1(aToReaderB, a);
                              return Curry._1(rToMB._0, r);
                            }));
              })
          };
  };
  var semiflatMap = function (aToMA, rToMA) {
    var rToMA$1 = rToMA._0;
    return /* ReaderT */{
            _0: (function (r) {
                return Curry._2(M.flat_map, Curry._1(rToMA$1, r), Curry.__1(aToMA));
              })
          };
  };
  var WithEnv = function (R) {
    var Functor = {
      map: map
    };
    var include = Relude__Relude_Extensions_Functor.FunctorExtensions(Functor);
    var Apply = {
      map: map,
      apply: apply
    };
    var include$1 = Relude__Relude_Extensions_Apply.ApplyExtensions(Apply);
    var Applicative = {
      map: map,
      apply: apply,
      pure: pure
    };
    var include$2 = Relude__Relude_Extensions_Applicative.ApplicativeExtensions(Applicative);
    var Monad = {
      map: map,
      apply: apply,
      pure: pure,
      flat_map: bind
    };
    var include$3 = Relude__Relude_Extensions_Monad.MonadExtensions(Monad);
    var include$4 = Relude__Relude_Extensions_Functor.FunctorInfix(Functor);
    var include$5 = Relude__Relude_Extensions_Apply.ApplyInfix(Apply);
    var include$6 = Relude__Relude_Extensions_Monad.MonadInfix(Monad);
    var Infix_FunctorExtensions = include$4.FunctorExtensions;
    var Infix_$less$$great = include$4.$less$$great;
    var Infix_$less$$$great = include$4.$less$$$great;
    var Infix_$less$ = include$4.$less$;
    var Infix_$$great = include$4.$$great;
    var Infix_$less$at$great = include$4.$less$at$great;
    var Infix_ApplyExtensions = include$5.ApplyExtensions;
    var Infix_$less$star$great = include$5.$less$star$great;
    var Infix_$less$star = include$5.$less$star;
    var Infix_$star$great = include$5.$star$great;
    var Infix_MonadExtensions = include$6.MonadExtensions;
    var Infix_$great$great$eq = include$6.$great$great$eq;
    var Infix_$eq$less$less = include$6.$eq$less$less;
    var Infix_$great$eq$great = include$6.$great$eq$great;
    var Infix_$less$eq$less = include$6.$less$eq$less;
    var Infix = {
      FunctorExtensions: Infix_FunctorExtensions,
      $less$$great: Infix_$less$$great,
      $less$$$great: Infix_$less$$$great,
      $less$: Infix_$less$,
      $$great: Infix_$$great,
      $less$at$great: Infix_$less$at$great,
      ApplyExtensions: Infix_ApplyExtensions,
      $less$star$great: Infix_$less$star$great,
      $less$star: Infix_$less$star,
      $star$great: Infix_$star$great,
      MonadExtensions: Infix_MonadExtensions,
      $great$great$eq: Infix_$great$great$eq,
      $eq$less$less: Infix_$eq$less$less,
      $great$eq$great: Infix_$great$eq$great,
      $less$eq$less: Infix_$less$eq$less
    };
    return {
            make: make,
            runReaderT: runReaderT,
            mapReaderT: mapReaderT,
            withReaderT: withReaderT,
            ask: ask,
            asks: asks,
            local: local,
            semiflatMap: semiflatMap,
            Functor: Functor,
            map: map,
            BsFunctorExtensions: include.BsFunctorExtensions,
            flipMap: include.flipMap,
            $$void: include.$$void,
            voidRight: include.voidRight,
            voidLeft: include.voidLeft,
            flap: include.flap,
            Apply: Apply,
            apply: apply,
            BsApplyExtensions: include$1.BsApplyExtensions,
            applyFirst: include$1.applyFirst,
            applySecond: include$1.applySecond,
            map2: include$1.map2,
            map3: include$1.map3,
            map4: include$1.map4,
            map5: include$1.map5,
            tuple2: include$1.tuple2,
            tuple3: include$1.tuple3,
            tuple4: include$1.tuple4,
            tuple5: include$1.tuple5,
            mapTuple2: include$1.mapTuple2,
            mapTuple3: include$1.mapTuple3,
            mapTuple4: include$1.mapTuple4,
            mapTuple5: include$1.mapTuple5,
            Applicative: Applicative,
            pure: pure,
            BsApplicativeExtensions: include$2.BsApplicativeExtensions,
            liftA1: include$2.liftA1,
            all: include$2.all,
            Monad: Monad,
            bind: bind,
            BsMonadExtensions: include$3.BsMonadExtensions,
            flatMap: include$3.flatMap,
            flatten: include$3.flatten,
            composeKleisli: include$3.composeKleisli,
            flipComposeKleisli: include$3.flipComposeKleisli,
            liftM1: include$3.liftM1,
            when_: include$3.when_,
            unless: include$3.unless,
            Infix: Infix
          };
  };
  var WithMonad = {
    make: make,
    runReaderT: runReaderT,
    mapReaderT: mapReaderT,
    withReaderT: withReaderT,
    ask: ask,
    asks: asks,
    local: local,
    map: map,
    apply: apply,
    pure: pure,
    bind: bind,
    semiflatMap: semiflatMap,
    WithEnv: WithEnv
  };
  var Functor = {
    map: map
  };
  var include = Relude__Relude_Extensions_Functor.FunctorExtensions(Functor);
  var Apply = {
    map: map,
    apply: apply
  };
  var include$1 = Relude__Relude_Extensions_Apply.ApplyExtensions(Apply);
  var Applicative = {
    map: map,
    apply: apply,
    pure: pure
  };
  var include$2 = Relude__Relude_Extensions_Applicative.ApplicativeExtensions(Applicative);
  var Monad = {
    map: map,
    apply: apply,
    pure: pure,
    flat_map: bind
  };
  var include$3 = Relude__Relude_Extensions_Monad.MonadExtensions(Monad);
  var include$4 = Relude__Relude_Extensions_Functor.FunctorInfix(Functor);
  var include$5 = Relude__Relude_Extensions_Apply.ApplyInfix(Apply);
  var include$6 = Relude__Relude_Extensions_Monad.MonadInfix(Monad);
  var Infix_FunctorExtensions = include$4.FunctorExtensions;
  var Infix_$less$$great = include$4.$less$$great;
  var Infix_$less$$$great = include$4.$less$$$great;
  var Infix_$less$ = include$4.$less$;
  var Infix_$$great = include$4.$$great;
  var Infix_$less$at$great = include$4.$less$at$great;
  var Infix_ApplyExtensions = include$5.ApplyExtensions;
  var Infix_$less$star$great = include$5.$less$star$great;
  var Infix_$less$star = include$5.$less$star;
  var Infix_$star$great = include$5.$star$great;
  var Infix_MonadExtensions = include$6.MonadExtensions;
  var Infix_$great$great$eq = include$6.$great$great$eq;
  var Infix_$eq$less$less = include$6.$eq$less$less;
  var Infix_$great$eq$great = include$6.$great$eq$great;
  var Infix_$less$eq$less = include$6.$less$eq$less;
  var Infix = {
    FunctorExtensions: Infix_FunctorExtensions,
    $less$$great: Infix_$less$$great,
    $less$$$great: Infix_$less$$$great,
    $less$: Infix_$less$,
    $$great: Infix_$$great,
    $less$at$great: Infix_$less$at$great,
    ApplyExtensions: Infix_ApplyExtensions,
    $less$star$great: Infix_$less$star$great,
    $less$star: Infix_$less$star,
    $star$great: Infix_$star$great,
    MonadExtensions: Infix_MonadExtensions,
    $great$great$eq: Infix_$great$great$eq,
    $eq$less$less: Infix_$eq$less$less,
    $great$eq$great: Infix_$great$eq$great,
    $less$eq$less: Infix_$less$eq$less
  };
  var include_BsFunctorExtensions = include.BsFunctorExtensions;
  var include_flipMap = include.flipMap;
  var include_void = include.$$void;
  var include_voidRight = include.voidRight;
  var include_voidLeft = include.voidLeft;
  var include_flap = include.flap;
  var include_BsApplyExtensions = include$1.BsApplyExtensions;
  var include_applyFirst = include$1.applyFirst;
  var include_applySecond = include$1.applySecond;
  var include_map2 = include$1.map2;
  var include_map3 = include$1.map3;
  var include_map4 = include$1.map4;
  var include_map5 = include$1.map5;
  var include_tuple2 = include$1.tuple2;
  var include_tuple3 = include$1.tuple3;
  var include_tuple4 = include$1.tuple4;
  var include_tuple5 = include$1.tuple5;
  var include_mapTuple2 = include$1.mapTuple2;
  var include_mapTuple3 = include$1.mapTuple3;
  var include_mapTuple4 = include$1.mapTuple4;
  var include_mapTuple5 = include$1.mapTuple5;
  var include_BsApplicativeExtensions = include$2.BsApplicativeExtensions;
  var include_liftA1 = include$2.liftA1;
  var include_all = include$2.all;
  var include_BsMonadExtensions = include$3.BsMonadExtensions;
  var include_flatMap = include$3.flatMap;
  var include_flatten = include$3.flatten;
  var include_composeKleisli = include$3.composeKleisli;
  var include_flipComposeKleisli = include$3.flipComposeKleisli;
  var include_liftM1 = include$3.liftM1;
  var include_when_ = include$3.when_;
  var include_unless = include$3.unless;
  return {
          WithMonad: WithMonad,
          make: make,
          runReaderT: runReaderT,
          mapReaderT: mapReaderT,
          withReaderT: withReaderT,
          ask: ask,
          asks: asks,
          local: local,
          semiflatMap: semiflatMap,
          Functor: Functor,
          map: map,
          BsFunctorExtensions: include_BsFunctorExtensions,
          flipMap: include_flipMap,
          $$void: include_void,
          voidRight: include_voidRight,
          voidLeft: include_voidLeft,
          flap: include_flap,
          Apply: Apply,
          apply: apply,
          BsApplyExtensions: include_BsApplyExtensions,
          applyFirst: include_applyFirst,
          applySecond: include_applySecond,
          map2: include_map2,
          map3: include_map3,
          map4: include_map4,
          map5: include_map5,
          tuple2: include_tuple2,
          tuple3: include_tuple3,
          tuple4: include_tuple4,
          tuple5: include_tuple5,
          mapTuple2: include_mapTuple2,
          mapTuple3: include_mapTuple3,
          mapTuple4: include_mapTuple4,
          mapTuple5: include_mapTuple5,
          Applicative: Applicative,
          pure: pure,
          BsApplicativeExtensions: include_BsApplicativeExtensions,
          liftA1: include_liftA1,
          all: include_all,
          Monad: Monad,
          bind: bind,
          BsMonadExtensions: include_BsMonadExtensions,
          flatMap: include_flatMap,
          flatten: include_flatten,
          composeKleisli: include_composeKleisli,
          flipComposeKleisli: include_flipComposeKleisli,
          liftM1: include_liftM1,
          when_: include_when_,
          unless: include_unless,
          Infix: Infix
        };
}

function make(rToMA) {
  return /* ReaderT */{
          _0: rToMA
        };
}

function runReaderT(r, rToMA) {
  return Curry._1(rToMA._0, r);
}

function mapReaderT(maToMB, rToMA) {
  var rToMA$1 = rToMA._0;
  return /* ReaderT */{
          _0: (function (param) {
              return Relude__Relude_Function.Infix.$less$less(maToMB, rToMA$1, param);
            })
        };
}

function withReaderT(r2ToR1, r1ToMA) {
  var r1ToMA$1 = r1ToMA._0;
  return /* ReaderT */{
          _0: (function (param) {
              return Relude__Relude_Function.Infix.$less$less(r1ToMA$1, r2ToR1, param);
            })
        };
}

var ask = /* ReaderT */{
  _0: (function (r) {
      return Curry._1(Relude__Relude_Identity.Monad.pure, r);
    })
};

function asks(rToA) {
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._1(Relude__Relude_Identity.Monad.pure, Curry._1(rToA, r));
            })
        };
}

function local(rToR, rToMA) {
  var rToMA$1 = rToMA._0;
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._1(rToMA$1, Curry._1(rToR, r));
            })
        };
}

function map(aToB, rToMA) {
  var rToMA$1 = rToMA._0;
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._2(Relude__Relude_Identity.Monad.map, aToB, Curry._1(rToMA$1, r));
            })
        };
}

function apply(rToMAToB, rToMA) {
  var rToMA$1 = rToMA._0;
  var rToMAToB$1 = rToMAToB._0;
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._2(Relude__Relude_Identity.Monad.apply, Curry._1(rToMAToB$1, r), Curry._1(rToMA$1, r));
            })
        };
}

function pure(a) {
  return /* ReaderT */{
          _0: (function (param) {
              return Curry._1(Relude__Relude_Identity.Monad.pure, a);
            })
        };
}

function bind(rToMA, aToReaderB) {
  var rToMA$1 = rToMA._0;
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._2(Relude__Relude_Identity.Monad.flat_map, Curry._1(rToMA$1, r), (function (a) {
                            var rToMB = Curry._1(aToReaderB, a);
                            return Curry._1(rToMB._0, r);
                          }));
            })
        };
}

function semiflatMap(aToMA, rToMA) {
  var rToMA$1 = rToMA._0;
  return /* ReaderT */{
          _0: (function (r) {
              return Curry._2(Relude__Relude_Identity.Monad.flat_map, Curry._1(rToMA$1, r), Curry.__1(aToMA));
            })
        };
}

function WithEnv(R) {
  var Functor = {
    map: map
  };
  var include = Relude__Relude_Extensions_Functor.FunctorExtensions(Functor);
  var Apply = {
    map: map,
    apply: apply
  };
  var include$1 = Relude__Relude_Extensions_Apply.ApplyExtensions(Apply);
  var Applicative = {
    map: map,
    apply: apply,
    pure: pure
  };
  var include$2 = Relude__Relude_Extensions_Applicative.ApplicativeExtensions(Applicative);
  var Monad = {
    map: map,
    apply: apply,
    pure: pure,
    flat_map: bind
  };
  var include$3 = Relude__Relude_Extensions_Monad.MonadExtensions(Monad);
  var include$4 = Relude__Relude_Extensions_Functor.FunctorInfix(Functor);
  var include$5 = Relude__Relude_Extensions_Apply.ApplyInfix(Apply);
  var include$6 = Relude__Relude_Extensions_Monad.MonadInfix(Monad);
  var Infix_FunctorExtensions = include$4.FunctorExtensions;
  var Infix_$less$$great = include$4.$less$$great;
  var Infix_$less$$$great = include$4.$less$$$great;
  var Infix_$less$ = include$4.$less$;
  var Infix_$$great = include$4.$$great;
  var Infix_$less$at$great = include$4.$less$at$great;
  var Infix_ApplyExtensions = include$5.ApplyExtensions;
  var Infix_$less$star$great = include$5.$less$star$great;
  var Infix_$less$star = include$5.$less$star;
  var Infix_$star$great = include$5.$star$great;
  var Infix_MonadExtensions = include$6.MonadExtensions;
  var Infix_$great$great$eq = include$6.$great$great$eq;
  var Infix_$eq$less$less = include$6.$eq$less$less;
  var Infix_$great$eq$great = include$6.$great$eq$great;
  var Infix_$less$eq$less = include$6.$less$eq$less;
  var Infix = {
    FunctorExtensions: Infix_FunctorExtensions,
    $less$$great: Infix_$less$$great,
    $less$$$great: Infix_$less$$$great,
    $less$: Infix_$less$,
    $$great: Infix_$$great,
    $less$at$great: Infix_$less$at$great,
    ApplyExtensions: Infix_ApplyExtensions,
    $less$star$great: Infix_$less$star$great,
    $less$star: Infix_$less$star,
    $star$great: Infix_$star$great,
    MonadExtensions: Infix_MonadExtensions,
    $great$great$eq: Infix_$great$great$eq,
    $eq$less$less: Infix_$eq$less$less,
    $great$eq$great: Infix_$great$eq$great,
    $less$eq$less: Infix_$less$eq$less
  };
  return {
          make: make,
          runReaderT: runReaderT,
          mapReaderT: mapReaderT,
          withReaderT: withReaderT,
          ask: ask,
          asks: asks,
          local: local,
          semiflatMap: semiflatMap,
          Functor: Functor,
          map: map,
          BsFunctorExtensions: include.BsFunctorExtensions,
          flipMap: include.flipMap,
          $$void: include.$$void,
          voidRight: include.voidRight,
          voidLeft: include.voidLeft,
          flap: include.flap,
          Apply: Apply,
          apply: apply,
          BsApplyExtensions: include$1.BsApplyExtensions,
          applyFirst: include$1.applyFirst,
          applySecond: include$1.applySecond,
          map2: include$1.map2,
          map3: include$1.map3,
          map4: include$1.map4,
          map5: include$1.map5,
          tuple2: include$1.tuple2,
          tuple3: include$1.tuple3,
          tuple4: include$1.tuple4,
          tuple5: include$1.tuple5,
          mapTuple2: include$1.mapTuple2,
          mapTuple3: include$1.mapTuple3,
          mapTuple4: include$1.mapTuple4,
          mapTuple5: include$1.mapTuple5,
          Applicative: Applicative,
          pure: pure,
          BsApplicativeExtensions: include$2.BsApplicativeExtensions,
          liftA1: include$2.liftA1,
          all: include$2.all,
          Monad: Monad,
          bind: bind,
          BsMonadExtensions: include$3.BsMonadExtensions,
          flatMap: include$3.flatMap,
          flatten: include$3.flatten,
          composeKleisli: include$3.composeKleisli,
          flipComposeKleisli: include$3.flipComposeKleisli,
          liftM1: include$3.liftM1,
          when_: include$3.when_,
          unless: include$3.unless,
          Infix: Infix
        };
}

var Reader = {
  make: make,
  runReaderT: runReaderT,
  mapReaderT: mapReaderT,
  withReaderT: withReaderT,
  ask: ask,
  asks: asks,
  local: local,
  map: map,
  apply: apply,
  pure: pure,
  bind: bind,
  semiflatMap: semiflatMap,
  WithEnv: WithEnv
};

exports.WithMonad = WithMonad;
exports.WithMonadAndEnv = WithMonadAndEnv;
exports.Reader = Reader;
/* Relude__Relude_Function Not a pure module */
