// relude/js/Relude_Js_Promise.js
// Generated by Melange
'use strict';

var Curry = require("melange.runtime/curry.js");
var Relude__Relude_IO = require("../Relude_IO.js");

function toIO(promise) {
  return Relude__Relude_IO.async(function (onDone) {
              promise.then(function (v) {
                      return Promise.resolve(Curry._1(onDone, {
                                      TAG: /* Ok */0,
                                      _0: v
                                    }));
                    }).catch(function (e) {
                    return Promise.resolve(Curry._1(onDone, {
                                    TAG: /* Error */1,
                                    _0: e
                                  }));
                  });
            });
}

function toIOLazy(runPromise) {
  return Relude__Relude_IO.async(function (onDone) {
              var promise = Curry._1(runPromise, undefined);
              promise.then(function (v) {
                      return Promise.resolve(Curry._1(onDone, {
                                      TAG: /* Ok */0,
                                      _0: v
                                    }));
                    }).catch(function (e) {
                    return Promise.resolve(Curry._1(onDone, {
                                    TAG: /* Error */1,
                                    _0: e
                                  }));
                  });
            });
}

function fromIOWithResult(io) {
  return new Promise((function (resolve, param) {
                Relude__Relude_IO.unsafeRunAsync((function (result) {
                        resolve(result);
                      }), io);
              }));
}

function fromIO(io) {
  return new Promise((function (resolve, reject) {
                Relude__Relude_IO.unsafeRunAsync((function (result) {
                        if (result.TAG === /* Ok */0) {
                          return resolve(result._0);
                        } else {
                          return reject(result._0);
                        }
                      }), io);
              }));
}

function fromIOExn(io) {
  return new Promise((function (resolve, reject) {
                Relude__Relude_IO.unsafeRunAsync((function (result) {
                        if (result.TAG === /* Ok */0) {
                          return resolve(result._0);
                        } else {
                          return reject(result._0);
                        }
                      }), io);
              }));
}

function fromIOJsExn(io) {
  return new Promise((function (resolve, reject) {
                Relude__Relude_IO.unsafeRunAsync((function (result) {
                        if (result.TAG === /* Ok */0) {
                          return resolve(result._0);
                        } else {
                          return reject(result._0);
                        }
                      }), io);
              }));
}

exports.toIO = toIO;
exports.toIOLazy = toIOLazy;
exports.fromIOWithResult = fromIOWithResult;
exports.fromIO = fromIO;
exports.fromIOExn = fromIOExn;
exports.fromIOJsExn = fromIOJsExn;
/* Relude__Relude_IO Not a pure module */
