// relude/js/Relude_Js_Console.js
// Generated by Melange
'use strict';

var Caml_splice_call = require("melange.runtime/caml_splice_call.js");
var Relude__Relude_IO = require("../Relude_IO.js");

function log(a) {
  return Relude__Relude_IO.suspend(function (param) {
              console.log(a);
            });
}

function log2(a, b) {
  return Relude__Relude_IO.suspend(function (param) {
              console.log(a, b);
            });
}

function log3(a, b, c) {
  return Relude__Relude_IO.suspend(function (param) {
              console.log(a, b, c);
            });
}

function log4(a, b, c, d) {
  return Relude__Relude_IO.suspend(function (param) {
              console.log(a, b, c, d);
            });
}

function logMany(xs) {
  return Relude__Relude_IO.suspend(function (param) {
              Caml_splice_call.spliceApply(console.log, [xs]);
            });
}

function warn(a) {
  return Relude__Relude_IO.suspend(function (param) {
              console.warn(a);
            });
}

var IO = {
  log: log,
  log2: log2,
  log3: log3,
  log4: log4,
  logMany: logMany,
  warn: warn
};

exports.IO = IO;
/* Relude__Relude_IO Not a pure module */
