// melange/bytesLabels.js
// Generated by Melange
'use strict';

var Stdlib__Bytes = require("./bytes.js");

var make = Stdlib__Bytes.make;

var init = Stdlib__Bytes.init;

var empty = Stdlib__Bytes.empty;

var copy = Stdlib__Bytes.copy;

var of_string = Stdlib__Bytes.of_string;

var to_string = Stdlib__Bytes.to_string;

var sub = Stdlib__Bytes.sub;

var sub_string = Stdlib__Bytes.sub_string;

var extend = Stdlib__Bytes.extend;

var fill = Stdlib__Bytes.fill;

var blit = Stdlib__Bytes.blit;

var blit_string = Stdlib__Bytes.blit_string;

var concat = Stdlib__Bytes.concat;

var cat = Stdlib__Bytes.cat;

var iter = Stdlib__Bytes.iter;

var iteri = Stdlib__Bytes.iteri;

var map = Stdlib__Bytes.map;

var mapi = Stdlib__Bytes.mapi;

var fold_left = Stdlib__Bytes.fold_left;

var fold_right = Stdlib__Bytes.fold_right;

var for_all = Stdlib__Bytes.for_all;

var exists = Stdlib__Bytes.exists;

var trim = Stdlib__Bytes.trim;

var escaped = Stdlib__Bytes.escaped;

var index = Stdlib__Bytes.index;

var index_opt = Stdlib__Bytes.index_opt;

var rindex = Stdlib__Bytes.rindex;

var rindex_opt = Stdlib__Bytes.rindex_opt;

var index_from = Stdlib__Bytes.index_from;

var index_from_opt = Stdlib__Bytes.index_from_opt;

var rindex_from = Stdlib__Bytes.rindex_from;

var rindex_from_opt = Stdlib__Bytes.rindex_from_opt;

var contains = Stdlib__Bytes.contains;

var contains_from = Stdlib__Bytes.contains_from;

var rcontains_from = Stdlib__Bytes.rcontains_from;

var uppercase = Stdlib__Bytes.uppercase;

var lowercase = Stdlib__Bytes.lowercase;

var capitalize = Stdlib__Bytes.capitalize;

var uncapitalize = Stdlib__Bytes.uncapitalize;

var uppercase_ascii = Stdlib__Bytes.uppercase_ascii;

var lowercase_ascii = Stdlib__Bytes.lowercase_ascii;

var capitalize_ascii = Stdlib__Bytes.capitalize_ascii;

var uncapitalize_ascii = Stdlib__Bytes.uncapitalize_ascii;

var compare = Stdlib__Bytes.compare;

var equal = Stdlib__Bytes.equal;

var starts_with = Stdlib__Bytes.starts_with;

var ends_with = Stdlib__Bytes.ends_with;

var unsafe_to_string = Stdlib__Bytes.unsafe_to_string;

var unsafe_of_string = Stdlib__Bytes.unsafe_of_string;

var split_on_char = Stdlib__Bytes.split_on_char;

var to_seq = Stdlib__Bytes.to_seq;

var to_seqi = Stdlib__Bytes.to_seqi;

var of_seq = Stdlib__Bytes.of_seq;

exports.make = make;
exports.init = init;
exports.empty = empty;
exports.copy = copy;
exports.of_string = of_string;
exports.to_string = to_string;
exports.sub = sub;
exports.sub_string = sub_string;
exports.extend = extend;
exports.fill = fill;
exports.blit = blit;
exports.blit_string = blit_string;
exports.concat = concat;
exports.cat = cat;
exports.iter = iter;
exports.iteri = iteri;
exports.map = map;
exports.mapi = mapi;
exports.fold_left = fold_left;
exports.fold_right = fold_right;
exports.for_all = for_all;
exports.exists = exists;
exports.trim = trim;
exports.escaped = escaped;
exports.index = index;
exports.index_opt = index_opt;
exports.rindex = rindex;
exports.rindex_opt = rindex_opt;
exports.index_from = index_from;
exports.index_from_opt = index_from_opt;
exports.rindex_from = rindex_from;
exports.rindex_from_opt = rindex_from_opt;
exports.contains = contains;
exports.contains_from = contains_from;
exports.rcontains_from = rcontains_from;
exports.uppercase = uppercase;
exports.lowercase = lowercase;
exports.capitalize = capitalize;
exports.uncapitalize = uncapitalize;
exports.uppercase_ascii = uppercase_ascii;
exports.lowercase_ascii = lowercase_ascii;
exports.capitalize_ascii = capitalize_ascii;
exports.uncapitalize_ascii = uncapitalize_ascii;
exports.compare = compare;
exports.equal = equal;
exports.starts_with = starts_with;
exports.ends_with = ends_with;
exports.unsafe_to_string = unsafe_to_string;
exports.unsafe_of_string = unsafe_of_string;
exports.split_on_char = split_on_char;
exports.to_seq = to_seq;
exports.to_seqi = to_seqi;
exports.of_seq = of_seq;
/* No side effect */
