// melange/pervasives.js
// Generated by Melange
'use strict';

var Stdlib = require("./stdlib.js");
var Caml_obj = require("melange.runtime/caml_obj.js");
var Caml_exceptions = require("melange.runtime/caml_exceptions.js");

var Exit = /* @__PURE__ */Caml_exceptions.create("Stdlib.Pervasives.Exit");

var min = Caml_obj.caml_min;

var max = Caml_obj.caml_max;

var nan = Number.NaN;

function $caret(prim0, prim1) {
  return prim0 + prim1;
}

function string_of_int(prim) {
  return String(prim);
}

function print_endline(prim) {
  console.log(prim);
}

function prerr_endline(prim) {
  console.error(prim);
}

var invalid_arg = Stdlib.invalid_arg;

var failwith = Stdlib.failwith;

var abs = Stdlib.abs;

var max_int = Stdlib.max_int;

var min_int = Stdlib.min_int;

var lnot = Stdlib.lnot;

var infinity = Stdlib.infinity;

var neg_infinity = Stdlib.neg_infinity;

var max_float = Stdlib.max_float;

var min_float = Stdlib.min_float;

var epsilon_float = Stdlib.epsilon_float;

var classify_float = Stdlib.classify_float;

var char_of_int = Stdlib.char_of_int;

var string_of_bool = Stdlib.string_of_bool;

var bool_of_string = Stdlib.bool_of_string;

var bool_of_string_opt = Stdlib.bool_of_string_opt;

var int_of_string_opt = Stdlib.int_of_string_opt;

var string_of_float = Stdlib.string_of_float;

var float_of_string_opt = Stdlib.float_of_string_opt;

var $at = Stdlib.$at;

var stdin = Stdlib.stdin;

var stdout = Stdlib.stdout;

var stderr = Stdlib.stderr;

var print_char = Stdlib.print_char;

var print_string = Stdlib.print_string;

var print_bytes = Stdlib.print_bytes;

var print_int = Stdlib.print_int;

var print_float = Stdlib.print_float;

var print_newline = Stdlib.print_newline;

var prerr_char = Stdlib.prerr_char;

var prerr_string = Stdlib.prerr_string;

var prerr_bytes = Stdlib.prerr_bytes;

var prerr_int = Stdlib.prerr_int;

var prerr_float = Stdlib.prerr_float;

var prerr_newline = Stdlib.prerr_newline;

var read_line = Stdlib.read_line;

var read_int = Stdlib.read_int;

var read_int_opt = Stdlib.read_int_opt;

var read_float = Stdlib.read_float;

var read_float_opt = Stdlib.read_float_opt;

var open_out = Stdlib.open_out;

var open_out_bin = Stdlib.open_out_bin;

var open_out_gen = Stdlib.open_out_gen;

var flush = Stdlib.flush;

var flush_all = Stdlib.flush_all;

var output_char = Stdlib.output_char;

var output_string = Stdlib.output_string;

var output_bytes = Stdlib.output_bytes;

var output = Stdlib.output;

var output_substring = Stdlib.output_substring;

var output_byte = Stdlib.output_byte;

var output_binary_int = Stdlib.output_binary_int;

var output_value = Stdlib.output_value;

var seek_out = Stdlib.seek_out;

var pos_out = Stdlib.pos_out;

var out_channel_length = Stdlib.out_channel_length;

var close_out = Stdlib.close_out;

var close_out_noerr = Stdlib.close_out_noerr;

var set_binary_mode_out = Stdlib.set_binary_mode_out;

var open_in = Stdlib.open_in;

var open_in_bin = Stdlib.open_in_bin;

var open_in_gen = Stdlib.open_in_gen;

var input_char = Stdlib.input_char;

var input_line = Stdlib.input_line;

var input = Stdlib.input;

var really_input = Stdlib.really_input;

var really_input_string = Stdlib.really_input_string;

var input_byte = Stdlib.input_byte;

var input_binary_int = Stdlib.input_binary_int;

var input_value = Stdlib.input_value;

var seek_in = Stdlib.seek_in;

var pos_in = Stdlib.pos_in;

var in_channel_length = Stdlib.in_channel_length;

var close_in = Stdlib.close_in;

var close_in_noerr = Stdlib.close_in_noerr;

var set_binary_mode_in = Stdlib.set_binary_mode_in;

var string_of_format = Stdlib.string_of_format;

var $caret$caret = Stdlib.$caret$caret;

var exit = Stdlib.exit;

var at_exit = Stdlib.at_exit;

var valid_float_lexem = Stdlib.valid_float_lexem;

var do_at_exit = Stdlib.do_at_exit;

exports.invalid_arg = invalid_arg;
exports.failwith = failwith;
exports.Exit = Exit;
exports.min = min;
exports.max = max;
exports.abs = abs;
exports.max_int = max_int;
exports.min_int = min_int;
exports.lnot = lnot;
exports.infinity = infinity;
exports.neg_infinity = neg_infinity;
exports.nan = nan;
exports.max_float = max_float;
exports.min_float = min_float;
exports.epsilon_float = epsilon_float;
exports.classify_float = classify_float;
exports.$caret = $caret;
exports.char_of_int = char_of_int;
exports.string_of_bool = string_of_bool;
exports.bool_of_string = bool_of_string;
exports.bool_of_string_opt = bool_of_string_opt;
exports.string_of_int = string_of_int;
exports.int_of_string_opt = int_of_string_opt;
exports.string_of_float = string_of_float;
exports.float_of_string_opt = float_of_string_opt;
exports.$at = $at;
exports.stdin = stdin;
exports.stdout = stdout;
exports.stderr = stderr;
exports.print_char = print_char;
exports.print_string = print_string;
exports.print_bytes = print_bytes;
exports.print_int = print_int;
exports.print_float = print_float;
exports.print_endline = print_endline;
exports.print_newline = print_newline;
exports.prerr_char = prerr_char;
exports.prerr_string = prerr_string;
exports.prerr_bytes = prerr_bytes;
exports.prerr_int = prerr_int;
exports.prerr_float = prerr_float;
exports.prerr_endline = prerr_endline;
exports.prerr_newline = prerr_newline;
exports.read_line = read_line;
exports.read_int = read_int;
exports.read_int_opt = read_int_opt;
exports.read_float = read_float;
exports.read_float_opt = read_float_opt;
exports.open_out = open_out;
exports.open_out_bin = open_out_bin;
exports.open_out_gen = open_out_gen;
exports.flush = flush;
exports.flush_all = flush_all;
exports.output_char = output_char;
exports.output_string = output_string;
exports.output_bytes = output_bytes;
exports.output = output;
exports.output_substring = output_substring;
exports.output_byte = output_byte;
exports.output_binary_int = output_binary_int;
exports.output_value = output_value;
exports.seek_out = seek_out;
exports.pos_out = pos_out;
exports.out_channel_length = out_channel_length;
exports.close_out = close_out;
exports.close_out_noerr = close_out_noerr;
exports.set_binary_mode_out = set_binary_mode_out;
exports.open_in = open_in;
exports.open_in_bin = open_in_bin;
exports.open_in_gen = open_in_gen;
exports.input_char = input_char;
exports.input_line = input_line;
exports.input = input;
exports.really_input = really_input;
exports.really_input_string = really_input_string;
exports.input_byte = input_byte;
exports.input_binary_int = input_binary_int;
exports.input_value = input_value;
exports.seek_in = seek_in;
exports.pos_in = pos_in;
exports.in_channel_length = in_channel_length;
exports.close_in = close_in;
exports.close_in_noerr = close_in_noerr;
exports.set_binary_mode_in = set_binary_mode_in;
exports.string_of_format = string_of_format;
exports.$caret$caret = $caret$caret;
exports.exit = exit;
exports.at_exit = at_exit;
exports.valid_float_lexem = valid_float_lexem;
exports.do_at_exit = do_at_exit;
/* nan Not a pure module */
