// melange/stringLabels.js
// Generated by Melange
'use strict';

var Stdlib__String = require("./string.js");

var make = Stdlib__String.make;

var init = Stdlib__String.init;

var empty = Stdlib__String.empty;

var of_bytes = Stdlib__String.of_bytes;

var to_bytes = Stdlib__String.to_bytes;

var concat = Stdlib__String.concat;

var cat = Stdlib__String.cat;

var equal = Stdlib__String.equal;

var compare = Stdlib__String.compare;

var starts_with = Stdlib__String.starts_with;

var ends_with = Stdlib__String.ends_with;

var contains_from = Stdlib__String.contains_from;

var rcontains_from = Stdlib__String.rcontains_from;

var contains = Stdlib__String.contains;

var sub = Stdlib__String.sub;

var split_on_char = Stdlib__String.split_on_char;

var map = Stdlib__String.map;

var mapi = Stdlib__String.mapi;

var fold_left = Stdlib__String.fold_left;

var fold_right = Stdlib__String.fold_right;

var for_all = Stdlib__String.for_all;

var exists = Stdlib__String.exists;

var trim = Stdlib__String.trim;

var escaped = Stdlib__String.escaped;

var uppercase_ascii = Stdlib__String.uppercase_ascii;

var lowercase_ascii = Stdlib__String.lowercase_ascii;

var capitalize_ascii = Stdlib__String.capitalize_ascii;

var uncapitalize_ascii = Stdlib__String.uncapitalize_ascii;

var iter = Stdlib__String.iter;

var iteri = Stdlib__String.iteri;

var index_from = Stdlib__String.index_from;

var index_from_opt = Stdlib__String.index_from_opt;

var rindex_from = Stdlib__String.rindex_from;

var rindex_from_opt = Stdlib__String.rindex_from_opt;

var index = Stdlib__String.index;

var index_opt = Stdlib__String.index_opt;

var rindex = Stdlib__String.rindex;

var rindex_opt = Stdlib__String.rindex_opt;

var to_seq = Stdlib__String.to_seq;

var to_seqi = Stdlib__String.to_seqi;

var of_seq = Stdlib__String.of_seq;

var blit = Stdlib__String.blit;

var copy = Stdlib__String.copy;

var fill = Stdlib__String.fill;

var uppercase = Stdlib__String.uppercase;

var lowercase = Stdlib__String.lowercase;

var capitalize = Stdlib__String.capitalize;

var uncapitalize = Stdlib__String.uncapitalize;

exports.make = make;
exports.init = init;
exports.empty = empty;
exports.of_bytes = of_bytes;
exports.to_bytes = to_bytes;
exports.concat = concat;
exports.cat = cat;
exports.equal = equal;
exports.compare = compare;
exports.starts_with = starts_with;
exports.ends_with = ends_with;
exports.contains_from = contains_from;
exports.rcontains_from = rcontains_from;
exports.contains = contains;
exports.sub = sub;
exports.split_on_char = split_on_char;
exports.map = map;
exports.mapi = mapi;
exports.fold_left = fold_left;
exports.fold_right = fold_right;
exports.for_all = for_all;
exports.exists = exists;
exports.trim = trim;
exports.escaped = escaped;
exports.uppercase_ascii = uppercase_ascii;
exports.lowercase_ascii = lowercase_ascii;
exports.capitalize_ascii = capitalize_ascii;
exports.uncapitalize_ascii = uncapitalize_ascii;
exports.iter = iter;
exports.iteri = iteri;
exports.index_from = index_from;
exports.index_from_opt = index_from_opt;
exports.rindex_from = rindex_from;
exports.rindex_from_opt = rindex_from_opt;
exports.index = index;
exports.index_opt = index_opt;
exports.rindex = rindex;
exports.rindex_opt = rindex_opt;
exports.to_seq = to_seq;
exports.to_seqi = to_seqi;
exports.of_seq = of_seq;
exports.blit = blit;
exports.copy = copy;
exports.fill = fill;
exports.uppercase = uppercase;
exports.lowercase = lowercase;
exports.capitalize = capitalize;
exports.uncapitalize = uncapitalize;
/* No side effect */
